Adding Particle System (port from AC)

DedeHai opened the pull request:
- port from AC with minor adaptations
- also adding hardware random functions



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a comprehensive particle system supporting both 2D and 1D LED animations with advanced physics and rendering.
  - Added over 30 new particle-based lighting effects including fire, volcano, fireworks, vortex, and more.
  - Expanded the total number of lighting modes to 228.
  - Integrated hardware-based true random number generation for improved randomness in effects.

- **Improvements**
  - Enhanced visual complexity with particle collision physics, customizable size control, motion and smear blur effects, and flexible particle emission settings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

DedeHai committed:
Adding Particle System (port from AC)

- port from AC
- also adding hardware random functions

DedeHai committed:
cleanup and minor improvements

- removed local buffer for ESP8266 in 1D system to save on RAM
- increased particle brightness in PS Impact
- minor tweak in collision binning (might improve speed)
- removed comments and some other unused stuff
- fixed a few compiler wranings

DedeHai committed:
fixed init sequence bug

netmindz commented on wled00/FX.h:
```diff
@@ -324,20 +324,50 @@ bool strip_uses_global_leds(void) __attribute__((pure));  // WLEDMM implemented
 #define FX_MODE_2DAKEMI                186
-#define FX_MODE_ARTIFX                 187 //WLEDMM ARTIFX
-#define FX_MODE_PARTYJERK              188
+#define FX_MODE_PARTICLEVOLCANO        187
```
Why have you changed the ID for existing effects? Can't your new ones just be added to the end?

DedeHai replied:
sure can. I copied that from here: https://github.com/MoonModules/WLED-MM/pull/234 and thought thats how you wanted it distributed.

netmindz replied:
Ah ok, my bad then 😅

With that I was trying to use the same commits to help preserve the attribution of the code and also simplify merging.

It was a bit of a rough quick test to see what challenges we may face rather than a serious PR

We should avoid changing the ID for existing effects as otherwise we break people's presets

DedeHai committed:
reverted FX re-enumeration

DedeHai committed:
fixed wrong gravity setting, added option for no trail (#4665)

DedeHai committed:
Bugfixes in PS, improvements to PS Fireworks 1D (#4673)

- fixed inconsitencies in size rendering
- fixed palette being wrapped in color by position and color by age modes
- Fixed bug in memory layout: for some unknown reason, if flags come before particles, last flag is sometimes overwritten, changing memory laout seems to fix that
- New color modes in PS Fireworks 1D:
 - custom3 slider < 16: lower saturation (check1: single color or multi-color explosions)
 - custom3 slider <= 23: full saturation (check1: single color or multi-color explosions)
 - custom3 slider > 23: color by speed (check 1 has not effect here)
 - custom slider = max: color by age or color by position (depends on check1)

DedeHai commented on wled00/FXparticleSystem.cpp:
```diff
@@ -0,0 +1,1941 @@
+    int32_t newY = part.y + (int32_t)part.vy;
+    partFlags.outofbounds = false; // reset out of bounds (in case particle was created outside the matrix and is now moving into view) note: moving this to checks below adds code and is not faster
+
+    if (advancedproperties) { //using individual particle size?
+      setParticleSize(particlesize); // updates default particleHardRadius
+      if (advancedproperties->size > PS_P_MINHARDRADIUS) {
+        particleHardRadius += (advancedproperties->size - PS_P_MINHARDRADIUS); // update radius
+        renderradius = particleHardRadius;
+      }
+    }
+    // note: if wall collisions are enabled, bounce them before they reach the edge, it looks much nicer if the particle does not go half out of view
```
this is intentional, not using the function leads to unnecessary code-duplication

DedeHai commented on wled00/FXparticleSystem.cpp:
```diff
@@ -0,0 +1,1925 @@
+      xsize = xsize > 64 ? xsize - 64 : 0;
+      ysize = ysize > 64 ? ysize - 64 : 0;
+    }
+
+    // calculate origin coordinates to render the particle to in the framebuffer
+    uint32_t xfb_orig = x - (rendersize>>1) + 1 - offset;
+    uint32_t yfb_orig = y - (rendersize>>1) + 1 - offset;
+    uint32_t xfb, yfb; // coordinates in frame buffer to write to note: by making this uint, only overflow has to be checked (spits a warning though)
+
+    //note on y-axis flip: WLED has the y-axis defined from top to bottom, so y coordinates must be flipped. doing this in the buffer xfer clashes with 1D/2D combined rendering, which does not invert y
+    //                     transferring the 1D buffer in inverted fashion will flip the x-axis of overlaid 2D FX, so the y-axis flip is done here so the buffer is flipped in y, giving correct results
+
+    // transfer particle renderbuffer to framebuffer
+    for (uint32_t xrb = offset; xrb < rendersize + offset; xrb++) {
+      xfb = xfb_orig + xrb;
+      if (xfb > (uint32_t)maxXpixel) {
+      if (wrapX) { // wrap x to the other side if required
+        if (xfb > (uint32_t)maxXpixel << 1) // xfb is "negative", handle it
+          xfb = (maxXpixel + 1) + (int32_t)xfb; // this always overflows to within bounds
+        else
+          xfb = xfb % (maxXpixel + 1); // note: without the above "negative" check, this works only for powers of 2
+      }
+      else
+        continue;
+      }
+
+      for (uint32_t yrb = offset; yrb < rendersize + offset; yrb++) {
+        yfb = yfb_orig + yrb;
+        if (yfb > (uint32_t)maxYpixel) {
+          if (wrapY) {// wrap y to the other side if required
+            if (yfb > (uint32_t)maxYpixel << 1) // yfb is "negative", handle it
+              yfb = (maxYpixel + 1) + (int32_t)yfb; // this always overflows to within bounds
+            else
+              yfb = yfb % (maxYpixel + 1); // note: without the above "negative" check, this works only for powers of 2
+          }
+          else
+          continue;
+        }
+        fast_color_add(framebuffer[xfb + (maxYpixel - yfb) * (maxXpixel + 1)], renderbuffer[xrb + yrb * 10]);
+      }
```
there is enough stack memory available, using heap causes heap fragmentation.

netmindz commented:
One blocker for merge is that PS isn't optional but merging in will break any builds that use animartrix as this would take us over 254 effects

DedeHai commented:
If I counted right, the currently used effect number is 220 (with PS effects).
https://github.com/wled/WLED/pull/4573 could be pulled into MM, removing 13 effect numbers by consolidating.
Animartrix uses 52 effect numbers, so without consolidating those the max number of standard effects can not be larger than 102.
this leaves us with 220-13 = 207, so 5 effects too many. 
If effects that have a PS replacement are removed (see `WLED_PS_DONT_REPLACE_FX` in AC) it removes another 9 effects from the list. 
since animartrix is 2D only: could also add `WLED_DISABLE_PARTICLESYSTEM1D` instead of `WLED_PS_DONT_REPLACE_FX` and remove the 15 1D PS effects. 
In the long run: Animartrix effects could be consolidated to save more IDs.

netmindz commented:
Given MoonModules focus on 2D and larger more complex setups Vs WLED, I'm much more comfortable ditching basic 1D legacy effects to free to resources for PS
